src/server/pkg/storage/fileset/storage.go

```diff
@@ -107,194 +99,148 @@ func (s *Storage) newReader(fileSet string, opts ...index.Option) *Reader {
+				return nil, err
+			}
+			fss = append(fss, fs)
+		}
 	}
 	if len(fss) == 0 {
-		return nil, errors.Errorf("error opening fileset: non-existent fileset: %v", fileSets)
+		return emptyFileSet{}, nil
 	}
 	if len(fss) == 1 {
 		return fss[0], nil
 	}
 	return newMergeReader(s.chunks, fss), nil
 }
 
-// Shard shards the file set into path ranges.
-// TODO This should be extended to be more configurable (different criteria
-// for creating shards).
-func (s *Storage) Shard(ctx context.Context, fs FileSet, cb ShardCallback) error {
-	return shard(ctx, fs, s.shardThreshold, cb)
+// Compose produces a composite fileset from the filesets under ids.
+// It does not perform a merge or check that the filesets at ids in any way
+// other than ensuring that they exist.
+func (s *Storage) Compose(ctx context.Context, ids []ID, ttl time.Duration) (*ID, error) {
+	c := &Composite{
+		Layers: ids,
+	}
+	return s.newComposite(ctx, c, ttl)
 }
 
-// ShardCallback is a callback that returns a path range for each shard.
-type ShardCallback func(*index.PathRange) error
+// Clone creates a new fileset, identical to the fileset at id, but with the specified ttl.
+// The ttl can be ignored by using track.NoTTL
+func (s *Storage) Clone(ctx context.Context, id ID, ttl time.Duration) (*ID, error) {
+	md, err := s.store.Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	switch x := md.Value.(type) {
+	case *Metadata_Primitive:
+		return s.newPrimitive(ctx, x.Primitive, ttl)
+	case *Metadata_Composite:
+		return s.newComposite(ctx, x.Composite, ttl)
+	default:
+		return nil, errors.Errorf("cannot clone type %T", md.Value)
+	}
+}
 
-// shard creates shards (path ranges) from the file set streams being merged.
-// A shard is created when the size of the content for a path range is greater than
-// the passed in shard threshold.
-// For each shard, the callback is called with the path range for the shard.
-func shard(ctx context.Context, fs FileSet, shardThreshold int64, cb ShardCallback) error {
-	var size int64
-	pathRange := &index.PathRange{}
-	if err := fs.Iterate(ctx, func(f File) error {
-		// A shard is created when we have encountered more than shardThreshold content bytes.
-		if size >= shardThreshold {
-			pathRange.Upper = f.Index().Path
-			if err := cb(pathRange); err != nil {
-				return err
-			}
-			size = 0
-			pathRange = &index.PathRange{
-				Lower: f.Index().Path,
+// Flatten takes a list of IDs and replaces references to composite FileSets
+// with references to all their layers inplace.
+// The returned IDs will only contain ids of Primitive FileSets
+func (s *Storage) Flatten(ctx context.Context, ids []ID) ([]ID, error) {
+	flattened := make([]ID, 0, len(ids))
+	for _, id := range ids {
+		md, err := s.store.Get(ctx, id)
+		if err != nil {
+			return nil, err
+		}
+		switch x := md.Value.(type) {
+		case *Metadata_Primitive:
+			flattened = append(flattened, id)
+		case *Metadata_Composite:
+			ids2, err := s.Flatten(ctx, x.Composite.Layers)
```

Comment:
Are you thinking we should support nesting composite filesets?

Reply:
Yeah, why not? I think we've talked about always compacting the flatten representation, but why do the extra work flattening if you don't have to.